src/core/lib/event_engine/windows/windows_engine.cc

```diff
@@ -140,15 +178,171 @@ bool WindowsEventEngine::IsWorkerThread() {
+  if (write_info->wsa_error() != 0) {
+    auto error = GRPC_WSA_ERROR(write_info->wsa_error(), "ConnectEx");
+    state->socket->MaybeShutdown(error);
+    state->on_connected_user_callback(error);
+    return;
+  }
+  // This code should be running in an executor thread already, so the callback
+  // can be run directly.
+  ChannelArgsEndpointConfig cfg;
+  state->on_connected_user_callback(std::make_unique<WindowsEndpoint>(
+      state->address, std::move(state->socket), std::move(state->allocator),
+      cfg, executor_.get()));
 }
 
 EventEngine::ConnectionHandle WindowsEventEngine::Connect(
     OnConnectCallback on_connect, const ResolvedAddress& addr,
-    const EndpointConfig& args, MemoryAllocator memory_allocator,
-    Duration deadline) {
-  GPR_ASSERT(false && "unimplemented");
+    const EndpointConfig& /* args */, MemoryAllocator memory_allocator,
+    Duration timeout) {
+  // TODO(hork): utilize the endpoint config
+  absl::Status status;
+  int istatus;
+  auto uri = ResolvedAddressToURI(addr);
+  if (!uri.ok()) {
+    Run([on_connect = std::move(on_connect), status = uri.status()]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  GRPC_EVENT_ENGINE_TRACE("EventEngine::%p connecting to %s", this,
+                          uri->c_str());
+  // Use dualstack sockets where available.
+  ResolvedAddress address = addr;
+  ResolvedAddress addr6_v4mapped;
+  if (ResolvedAddressToV4Mapped(addr, &addr6_v4mapped)) {
+    address = addr6_v4mapped;
+  }
+  SOCKET sock = WSASocket(AF_INET6, SOCK_STREAM, IPPROTO_TCP, nullptr, 0,
+                          IOCP::GetDefaultSocketFlags());
+  if (sock == INVALID_SOCKET) {
+    Run([on_connect = std::move(on_connect),
+         status = GRPC_WSA_ERROR(WSAGetLastError(), "WSASocket")]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  status = PrepareSocket(sock);
+  if (!status.ok()) {
+    Run([on_connect = std::move(on_connect), status]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  // Grab the function pointer for ConnectEx for that specific socket It may
+  // change depending on the interface.
+  LPFN_CONNECTEX ConnectEx;
+  GUID guid = WSAID_CONNECTEX;
+  DWORD ioctl_num_bytes;
+  istatus = WSAIoctl(sock, SIO_GET_EXTENSION_FUNCTION_POINTER, &guid,
+                     sizeof(guid), &ConnectEx, sizeof(ConnectEx),
+                     &ioctl_num_bytes, nullptr, nullptr);
+  if (istatus != 0) {
+    Run([on_connect = std::move(on_connect),
+         status = GRPC_WSA_ERROR(
+             WSAGetLastError(),
+             "WSAIoctl(SIO_GET_EXTENSION_FUNCTION_POINTER)")]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  // bind the local address
+  auto local_address = ResolvedAddressMakeWild6(0);
+  istatus = bind(sock, local_address.address(), local_address.size());
```

Comment:
why is there a bind to local address ? Could you add a comment

Reply:
https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-bind ... this would be the equivalent to adding a link to `man 2 bind` around the posix implementation's use of bind, it's a bit overkill.

> The bind function is required before the use of the [send](https://learn.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-send) or [WSASend](https://learn.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-wsasend) functions which do not perform an implicit bind and are allowed only on connected sockets, which means the socket must have already been bound for it to be connected.